カレンダープログラムを作成しました

sht3ng opened the pull request:
ご確認どうかよろしくお願いいたします。

nagano-shunta committed:
カレンダーを表示するためのプログラムを追加

必須要件と歓迎要件は満たしているが，メソッドやクラスにまとめていない段階

nagano-shunta committed:
カレンダー表示プログラムをメソッドにまとめました

再利用可能性を高めるために，ただの処理のかたまりだったカレンダープログラムをメソッドにまとめました。

nagano-shunta committed:
カレンダープログラムのメソッドと変数をclass Calendarにまとめました

散らばっていたメソッドと変数をまとめるためにclass Calendarを作成しました。

naganoshunta commented:
このプルリクエストには3つコミットがあります。
一つ目のコミットはメソッドに処理をまとめず，処理のまとまりをべた書きしている段階で，
二つ目のコミットは処理のまとまりをメソッドにまとめた段階で，
三つ目のコミットはメソッドと変数をクラスにまとめた段階のものです。

処理のまとまりをメソッドにまとめるときはどこからどこまでを一つのメソッドにするかが悩ましく，
メソッドと変数をクラスにまとめるときはどのようにクラスにまとめるかがとても悩ましかったです。
おかしなところもあると思いますが，ご確認のほどよろしくお願いいたします。

nagano-shunta committed:
mark_todayメソッドをeachを使わない方法に変更しました

mark_todayメソッドはeachを使わずに実現できることに気づいたため無駄なeachの使用を避けるコードに変更しました。

nagano-shunta committed:
get_tableメソッドとget_outputメソッドをprivateメソッドにしてinitialize時に呼び出すようにしました

get_tableよりも先にget_outputメソッドが呼ばれた場合，エラーが発生してしまう問題や，
get_outputメソッドよりも先にprint_outputメソッドが呼ばれた場合，望んだ結果が得られない問題を
解消するために，initialize時にget_tableメソッドとget_outputメソッドを呼び出し，
インスタンス変数@tableと@outputを作成するように変更しました。
また，この変更に伴い，get_tableメソッドとget_outputメソッドを外部から呼び出す必要がなくなったため，
これらのメソッドをprivateメソッドに変更しました。

nagano-shunta committed:
get_outputメソッドの可読性が低い問題を修正

無理に細かいメソッドに分けていたことで，get_outputメソッドの可読性が低かった問題を修正するために，
get_outputメソッド内で出力用のインスタンス変数@outputの整形を行うように変更しました。

nagano-shunta committed:
get_titleメソッドの行が横に長かった問題を修正しました

これまでget_titleメソッド内で+演算子による文字列の連結でカレンダーのタイトルを返していましたが，
行方向に長くなっていて見づらかったので，Array#joinによる文字列の連結に変更しました。

nagano-shunta committed:
is_saturday?とis_today?を定義してget_output内の条件分岐を判断させるようにしました

get_output内のcaseによる条件分岐で土曜かどうかと今日かどうかを確認するコードが
重複していたので，メソッドにして変更が生じたときにも変更しやすいようにしました。

nagano-shunta committed:
コマンドラインオプションの処理をクラスにまとめました

些細なことですがローカル変数の数を減らすことと処理のかたまりをクラスに整理するために，コマンドラインオプションの処理をクラスにまとめました。

sht3ng commented:
レビュワーの方へ
「naganoshunta force-pushed the calendar  branch from 832a4f9 to 0739f91」の履歴についてなのですが，
[02.calendar/calendar.rb の42行目](https://github.com/naganoshunta/ruby-practices/blob/0739f913d8015325e270eb5b035e7d1d802ef244/02.calendar/calendar.rb#L42)の`private`の下に改行を入れただけの変更をVSCodeのソース管理機能から`$ git commit --amend`した結果，Merge Commitが作成されてしまい，こちらを`$ git reset --hard 0739f91` と `$ git push --force-with-leace`コマンドによって取り消そうとした履歴です。
こちらの履歴によって，[02.calendar/calendar.rb の42行目](https://github.com/naganoshunta/ruby-practices/blob/0739f913d8015325e270eb5b035e7d1d802ef244/02.calendar/calendar.rb#L42)の`private`の下に改行を入れた以外にはコードに変更は加えられていません。
コミット履歴を汚してしまって大変申し訳ありません。

naganoshunta commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,125 @@
+    print @output
+  end
+
+  private
```
レビュワーの方へ
「naganoshunta force-pushed the calendar  branch from 832a4f9 to 0739f91」の履歴につきまして，
02.calendar/calendar.rb の42行目`private`の下に改行を入れただけの変更をVSCodeからgit commit --amendした結果，
Merge Commitが作成されてしまいこちらをgit reset --hard 0739f91 && git push --force-with-leaceコマンドによって
取り消そうとした履歴です。こちらの履歴によって，改行以外にはコード自体に変更は加えられていません。
コミット履歴が汚れてしまって大変申し訳ありませんが，よろしくお願いいたします。

nagano-shunta committed:
コマンドラインオプションで整数以外が与えられた場合の例外処理をOptionsクラスで行うように変更しました

これまでは整数以外がコマンドラインオプションで与えられた場合，
オプションのパース時にto_iメソッドによって整数へ変換を試みた後に，
Calendarクラス内のvalid_date?メソッドにて年と月として妥当かどうかを判断させていたのですが，
無効なオプションへの例外処理があいまいに取り扱われていた問題を解決するために変更を行いました。
また，"--month [MONTH]"と"--year [YEAR]"のように，[]をつけることでオプションの引数が必須ではないと
指定していたのですが，この指定は適切ではなかったという問題を解決するために変更を行いました。

nagano-shunta committed:
OptionParser#onに正規表現を追加して正しい年と月の指定以外で例外を発生させるようにしました

これまで年や月として正しくない整数がオプションの引数に渡されたときは，
Calendarクラスのvalid_date?で年と月として有効な整数かを判断させていたのですが，
オプションの引数の妥当性はオプションを管理するクラスに判断させた方が
よりシンプルだと考えたため，この変更を行いました。

nagano-shunta committed:
正規表現に一部冗長な箇所があったので修正しました

nagano-shunta committed:
スタイルを整えました（attr_reader後に改行追加，メソッド名の助動詞isを削除，.joinの改行を削除）

スタイルが統一されていない箇所（.joinの改行）や読みにくい箇所（attr_reader後に改行がない，メソッド名の冗長なis）が
含まれていた問題を修正するために変更しました。

nagano-shunta committed:
一部のメソッド名とインスタンス変数名を適切なものに変更しました

rubocop-fjordを走らせた結果，メソッド名に使うべきではない単語（get）を使っていたことに気づいたので，
メソッド名をより適切なもの（parseやgenerate）に変更しました。
式展開していないダブルクォーテーションをシングルクォーテーションに変更しました。
最終的に出力するカレンダーを作成するメソッドの複雑さを減らすために
@date_table変数に対して整形を行う方法に変更しました。

u1tnk commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,120 @@
+
+  def parse_options
+    opts = OptionParser.new
+    opts.on('-m', '--month MONTH', /(^[1-9]$)|(^1[0-2]$)/, Integer) do |month|
```
こだわりの正規表現ですね 👍

sht3ng replied:
ありがとうございます！☺️
拙いですがVSCodeの検索機能で試しながら作ってみました！

u1tnk commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,120 @@
+end
+
+class Calendar
+  attr_reader :today, :validity, :first_day, :last_day, :year, :month, :date_table, :calendar
```
このプラクティスだと気にする必要は無いのですが、オブジェクト指向的には外部からアクセスされないものには attr_readerは付けない方が良いですよ。

sht3ng replied:
ありがとうございます！とても勉強になります！😄
なんとなくつくったインスタンス変数はattr_readerなどに書いておかないといけないのかなと思い込んでいました！
使わないものはちゃんとしまっておかないといけないのですね！

u1tnk commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,120 @@
+  end
+
+  def generate_calendar
+    @calendar = [
```
Calendarクラスに `@calendar` 変数があるのはちょっと混乱してしまいますね。
また、変数名を変更するより、outputから呼び出されたときにこの文字列を返すようにすれば良いと思います。
一時的に保存する必要の無いものは呼び出されたときに生成すると良いですよ。

sht3ng replied:
こちらもとても勉強になりました！☺️
出力用の文字列は保存する必要ないものですね！
自分でも気づかないうちに変数に代入することに考えが固着していたように思います！

u1tnk commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,120 @@
+  end
+
+  def invert_color(string)
+    "\e[7m#{string}\e[0m"
```
歓迎要件 👍

sht3ng replied:
ありがとうございます！☺️

u1tnk reviewed (changes requested):
もうほとんどオブジェクト指向ですね 👍 
プラクティスの求めるレベル的にもほとんど良いのですが、誤解を招く変数名は良くないので、修正お願いします。

nagano-shunta committed:
レビューしていただいた内容を元にattr_readerで指定するインスタンス変数とカレンダーを出力するメソッドを変更しました

1. 外部からアクセスされないインスタンス変数にattr_readerをつけていた問題を修正するためにattr_readerを修正しました。
2. 一時的に保存する必要のない出力用カレンダーの文字列を@calendarに保存していた問題を修正するために，outputメソッド内で出力用の文字列を生成して戻り値とするように変更しました。

sht3ng commented:
@u1tnk さん
レビューしてくださりありがとうございました！
とても勉強になりました！😄
レビューいただいた内容をもとにコードを修正したコミットをpushいたしました！
今回の変更で出力用の文字列をoutputメソッド内で生成して戻り値にするように変更して外部でprintメソッドに渡すようにしたのですが，
printメソッドは外部で呼び出す形で問題なかったでしょうか。
お手数おかけして恐縮ですがご確認のほどどうかよろしくお願いいたします🙇‍♂️

u1tnk reviewed (approved):
色々スッキリしましたね 👍 

> printメソッドは外部で呼び出す形で問題なかったでしょうか。

はい、そちらの方がクラスの責務が減って良いと思いますよ。
ただ、 `output`というメソッドで文字列が返ってくるのもアリなレベルですが、 `to_s`とか`format`とかの方が分かりやすいかと思います。

そこは修正はお任せできるレベルなのでapproveです 🙆‍♀️

nagano-shunta committed:
レビューしていただいた内容を元にメソッド名の変更（output -> format）と@validity変数の削除を行いました

出力用のカレンダーの文字列を返すメソッドoutputの名前をより適切なformatという名前に変更いたしました。
また，@validity変数を使用していないことに気づいたため，こちらの変数を削除いたしました。

sht3ng commented:
レビューありがとうございます！☺️
基本的にクラスの責任はできるだけ小さくする方がよいのですね！勉強になりました！

> outputというメソッドで文字列が返ってくるのもアリなレベルですが、 to_sとかformatとかの方が分かりやすいかと思います。

どんなクラスのオブジェクトが戻り値として返ってくるかも考慮してメソッド名を考える必要があるのですね！
こちらも修正させていただきました！

お時間をとってみていただいてありがとうございました！☺️🙇‍♂️
